File: api/vetfunc.py
import json
from func import Funcionario

class VetorFuncionario:
    def __init__(self):
        self._funcionarios = [None] * 10
        self._qtsFunc = 0
        self._onde = 0
    
    def lerDeArquivo(self, nomeArquivo):
        try:
            arquivo = open(nomeArquivo, 'r', encoding='utf-8')
            dados = json.load(arquivo)
            self._funcionarios = [None] * 10
            self._qtsFunc = 0
            self._onde = 0
            for i in range(len(dados)):
                f = Funcionario()
                f.fromDict(dados[i])
                self._funcionarios[self._qtsFunc] = f
                self._qtsFunc += 1
            arquivo.close()
        except FileNotFoundError:
            print('Arquivo não encontrado')
        except Exception as erro:
            print(f'Ocorreu um erro: {erro}')
    
    def salvarEmArquivo(self, nomeArquivo):
        try:
            arquivo = open(nomeArquivo, 'w', encoding='utf-8')
            lista = [None] * len(self._funcionarios)
            qts = 0
            for i in range(self._qtsFunc):
                lista[qts] = self._funcionarios[i].toDict()
                qts += 1
            json.dump(lista, arquivo)
            arquivo.close()
        except Exception as erro:
            print(f'Ocorreu um erro: {erro}')
    
    def expandirVetor(self):
        novo = [None] * (len(self._funcionarios) * 2)
        for i in range(self._qtsFunc):
            novo[i] = self._funcionarios[i]
        self._funcionarios = novo

    def existeFuncionario(self, id):
        inicio = 0
        fim = self._qtsFunc - 1
        while inicio <= fim:
            self._onde = (inicio + fim) // 2
            if self._funcionarios[self._onde].id == id:
                return True
            if self._funcionarios[self._onde].id < id:
                inicio = self._onde + 1
            else:
                fim = self._onde - 1
        
    def adicionarFuncionario(self, func):
        if (not self.existeFuncionario(func.id)):
            if self._qtsFunc == len(self._funcionarios):
                self.expandirVetor()
            self._funcionarios[self._qtsFunc] = func
            self._qtsFunc += 1
    
    def removerFuncionario(self, id):
        if self.existeFuncionario(id):
            for i in range(self._onde, self._qtsFunc - 1):
                self._funcionarios[i] = self._funcionarios[i + 1]
            self._qtsFunc -= 1
    
    def buscarFuncionario(self, id):
        if self.existeFuncionario(id):
            return self._funcionarios[self._onde]
        return None

    @property
    def funcionarios(self):
        return self._funcionarios
    
    @funcionarios.setter
    def funcionarios(self, funcionarios):
        self._funcionarios = funcionarios

    @property
    def qtsFunc(self):
        return self._qtsFunc
    
    @qtsFunc.setter
    def qtsFunc(self, qtsFunc):
        self._qtsFunc = qtsFunc

    @property
    def onde(self):
        return self._onde
    
    @onde.setter
    def onde(self, onde):
        self._onde = onde

File: api/javascript/vetfunc.js
const Funcionario = require('./func.js');
const fs = require('fs');

class VetorFuncionario {
    constructor() {
        this._funcionarios = Array.apply(null, Array(10)).map(function () {});
        this._qtsFunc = 0;
        this._onde = 0;
    }

    lerDeArquivo(nomeArquivo) {
        try {
            fs.readFile(nomeArquivo, 'utf-8', (err, data) => {
                let dados = JSON.parse(data);
                this._funcionarios = Array.apply(null, Array(10)).map(function () {});
                this._qtsFunc = 0;
                this._onde = 0;
                for (let i = 0; i < dados.length; i++) {
                    if (this._qtsFunc == this._funcionarios.length) {
                        this.expandirVetor();
                    }
                    let f = new Funcionario();
                    f.fromJSON(dados[i]);
                    this._funcionarios[this._qtsFunc] = f;
                    this._qtsFunc++;
                }
            });
        } catch (e) {
            console.log(e);
        }
    }

    salvarEmArquivo(nomeArquivo) {
        try {
            let lista = Array.apply(null, Array(this._qtsFunc)).map(function () {});
            for (let i = 0; i < this._qtsFunc; i++) {
                lista[i] = this._funcionarios[i].toJSON();
            }
            console.log(lista);
            fs.writeFileSync(nomeArquivo, JSON.stringify(lista, null, 4), 'utf-8');
        } catch (e) {
            console.log(e);
        }
    }

    expandirVetor() {
        let novo = Array.apply(null, Array(this._funcionarios.length * 2)).map(function () {});
        for (let i = 0; i < this._funcionarios.length; i++) {
            novo[i] = this._funcionarios[i];
        }
        this._funcionarios = novo;
    }

    existeFuncionario(id) {
        let inicio = 0;
        let fim = this._qtsFunc - 1;
        while (inicio <= fim) {
            this._onde = Math.floor((inicio + fim) / 2);
            if (this._funcionarios[this._onde].id == id) {
                return true;
            }
            if (this._funcionarios[this._onde].id < id) {
                inicio = this._onde + 1;
            } else {
                fim = this._onde - 1;
            }
        }
        return false;
    }

    adicionarFuncionario(func) {
        if (!this.existeFuncionario(func.id)) {
            if (this._qtsFunc == this._funcionarios.length) {
                this.expandirVetor();
            }
            this._funcionarios[this._qtsFunc] = func;
            this._qtsFunc++;
        }
    }

    removerFuncionario(id) {
        if (this.existeFuncionario(id)) {
            console.log(this._funcionarios[this._onde]);
            for (let i = this._onde; i < this._qtsFunc - 1; i++) {
                this._funcionarios[i] = this._funcionarios[i + 1];
            }
            this._qtsFunc--;
        }
    }

    buscarFuncionario(id) {
        if (this.existeFuncionario(id)) {
            return this._funcionarios[this._onde];
        }
        return null;
    }

    get funcionarios() {
        return this._funcionarios;
    }

    set funcionarios(funcionarios) {
        this._funcionarios = funcionarios;
    }

    get qtsFunc() {
        return this._qtsFunc;
    }

    set qtsFunc(qtsFunc) {
        this._qtsFunc = qtsFunc;
    }

    get onde() {
        return this._onde;
    }

    set onde(onde) {
        this._onde = onde;
    }
}

module.exports = VetorFuncionario;